### PR TITLE
Guest panic message

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
Expose guest panic message in `GuestPanic` error.

Slightly modifying ABORT message format to avoid escaping strings.
Introduce `panic_utf8` API in runtime to throw a new message.
Fixed underflow issue with AS abort function.
Created TODO #1419  for `log_utf16`

More error checks going to be used in #1408

### CODE REVIEW

**Comment by evgenykuzyakov** [comment]
Nearlib change for this PR is pending: https://github.com/nearprotocol/nearlib/pull/85

File: runtime/near-vm-errors/src/lib.rs [hunk 1]
```diff
@@ -174,7 +174,7 @@ impl std::fmt::Display for HostError {
             GasLimitExceeded => write!(f, "Exceeded the maximum amount of gas allowed to burn per contract."),
             BalanceExceeded => write!(f, "Exceeded the account balance."),
             EmptyMethodName => write!(f, "Tried to call an empty method name."),
-            GuestPanic => write!(f, "Smart contract has explicitly invoked `panic`."),
+            GuestPanic(s) => write!(f, "Smart contract panicked: {}", s),
```
Comment (mikhailOK) [inline runtime/near-vm-errors/src/lib.rs]: should we escape the msg?

File: runtime/near-vm-logic/src/logic.rs [hunk 2]
```diff
@@ -264,6 +264,66 @@ impl<'a> VMLogic<'a> {
...
+            if len > self.config.max_log_len {
+                return Err(HostError::BadUTF8.into());
+            }
+            buf = Self::memory_get(self.memory, ptr, len)?;
+        } else {
+            buf = vec![];
+            for i in 0..=self.config.max_log_len {
+                if i == self.config.max_log_len {
+                    return Err(HostError::BadUTF8.into());
+                }
```
Comment (mikhailOK) [inline runtime/near-vm-logic/src/logic.rs]: `\0` counts towards max_log_len?

File: runtime/near-vm-logic/src/logic.rs [hunk 3]
```diff
@@ -264,6 +264,66 @@ impl<'a> VMLogic<'a> {
...
+    // # String reading helper functions #
+    // ###################################
+
+    /// Helper function to read and return utf8-encoding string.
+    /// If `len == u64::MAX` then treats the string as null-terminated with character `'\0'`.
+    ///
+    /// # Errors
+    ///
+    /// * If string extends outside the memory of the guest with `MemoryAccessViolation`;
+    /// * If string is not UTF-8 returns `BadUtf8`.
```
Comment (mikhailOK) [inline runtime/near-vm-logic/src/logic.rs]: if string is longer than `max_log_len` returns `BadUtf8`

File: runtime/near-vm-errors/src/lib.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply runtime/near-vm-errors/src/lib.rs]: No. I think we should refactor FunctionCallError to be an enum with typed errors. And panic with a message should be one of them .

File: runtime/near-vm-logic/src/logic.rs (hunk 2 above)
Comment (evgenykuzyakov) [reply runtime/near-vm-logic/src/logic.rs]: Good point. Moved the check below null-termination check.

**Pull request merged**